File: src/ai_agent/data/thirteenf.py
"""SEC EDGAR 13F holdings — institutional "smart money" fetcher.

Pulls a fund manager's most recent 13F-HR filing (the quarterly disclosure of
long US-equity holdings) from EDGAR, so the agent can weigh what widely-followed
investors hold. Mirrors ``lib/thirteenf.ts`` (the /leaders page).

13F data only changes once a quarter, so successful results are cached for the
process lifetime.
"""

from __future__ import annotations

import logging
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field

import httpx

logger = logging.getLogger(__name__)

_TIMEOUT = httpx.Timeout(15.0)
_SUBMISSIONS = "https://data.sec.gov/submissions/CIK{cik}.json"
_ARCHIVE = "https://www.sec.gov/Archives/edgar/data/{cik}/{acc}"


@dataclass(frozen=True)
class Manager:
    name: str
    cik: str  # 10-digit zero-padded


@dataclass(frozen=True)
class Holding:
    issuer: str
    cusip: str
    value: float  # USD, as reported in the filing
    pct: float  # share of the manager's 13F portfolio (0..1)


@dataclass
class Report:
    manager: str
    cik: str
    period_of_report: str | None = None
    holdings: list[Holding] = field(default_factory=list)
    error: str | None = None


# Widely-followed institutional managers, by SEC CIK.
MANAGERS: list[Manager] = [
    Manager("Berkshire Hathaway — Warren Buffett", "0001067983"),
    Manager("Scion Asset Management — Michael Burry", "0001649339"),
    Manager("Pershing Square — Bill Ackman", "0001336528"),
]

# Process-lifetime cache: CIK -> Report (13F data only changes quarterly).
_cache: dict[str, Report] = {}


def _user_agent() -> str:
    from ai_agent.settings import get_settings

    return get_settings().edgar_user_agent


def _localname(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def latest_13f(manager: Manager, *, client: httpx.Client | None = None) -> Report:
    """Return *manager*'s latest 13F-HR holdings; cached per process on success."""
    if manager.cik in _cache:
        return _cache[manager.cik]
    report = _fetch(manager, client)
    if report.error is None:
        _cache[manager.cik] = report
    return report


def _fetch(manager: Manager, client: httpx.Client | None) -> Report:
    own = client is None
    cl = client or httpx.Client(timeout=_TIMEOUT, headers={"User-Agent": _user_agent()})
    try:
        subs = cl.get(_SUBMISSIONS.format(cik=manager.cik))
        if subs.status_code != 200:
            return Report(manager.name, manager.cik, error=f"submissions HTTP {subs.status_code}")
        recent = (subs.json().get("filings") or {}).get("recent") or {}
        forms = recent.get("form") or []
        idx = next((i for i, f in enumerate(forms) if f in ("13F-HR", "13F-HR/A")), None)
        if idx is None:
            return Report(manager.name, manager.cik, error="no 13F-HR filing found")

        accession = (recent.get("accessionNumber") or [])[idx]
        period = (recent.get("reportDate") or [])[idx] if recent.get("reportDate") else None
        base = _ARCHIVE.format(cik=str(int(manager.cik)), acc=accession.replace("-", ""))

        index = cl.get(f"{base}/index.json")
        if index.status_code != 200:
            return Report(
                manager.name, manager.cik, period, error=f"index HTTP {index.status_code}"
            )
        items = (index.json().get("directory") or {}).get("item") or []
        info_name = next(
            (
                it["name"]
                for it in items
                if isinstance(it.get("name"), str)
                and "info" in it["name"].lower()
                and "table" in it["name"].lower()
                and it["name"].lower().endswith(".xml")
            ),
            None,
        )
        if not info_name:
            return Report(manager.name, manager.cik, period, error="no info table in filing")

        xml = cl.get(f"{base}/{info_name}")
        if xml.status_code != 200:
            return Report(
                manager.name, manager.cik, period, error=f"info table HTTP {xml.status_code}"
            )
        return Report(manager.name, manager.cik, period, holdings=_parse_holdings(xml.text))
    except Exception as exc:  # network failure, XML parse error, …
        logger.warning("13F fetch failed for %s: %s", manager.name, exc)
        return Report(manager.name, manager.cik, error=str(exc))
    finally:
        if own:
            cl.close()


def _parse_holdings(xml: str) -> list[Holding]:
    """Parse a 13F information-table XML into merged, value-ranked holdings."""
    # Encode to bytes: ElementTree rejects str input carrying an XML encoding
    # declaration, which EDGAR's information tables always include.
    root = ET.fromstring(xml.encode("utf-8"))
    merged: dict[str, dict] = {}
    for el in root.iter():
        if _localname(el.tag) != "infoTable":
            continue
        fields: dict[str, str] = {}
        for child in el.iter():
            if child.text and child.text.strip():
                fields[_localname(child.tag)] = child.text.strip()
        try:
            value = float(fields.get("value", "0"))
        except ValueError:
            continue
        if value <= 0:
            continue
        issuer = fields.get("nameOfIssuer", "—")
        cusip = fields.get("cusip", "")
        key = cusip or issuer
        if key in merged:
            merged[key]["value"] += value
        else:
            merged[key] = {"issuer": issuer, "cusip": cusip, "value": value}

    total = sum(h["value"] for h in merged.values())
    holdings = [
        Holding(
            issuer=h["issuer"],
            cusip=h["cusip"],
            value=h["value"],
            pct=(h["value"] / total) if total > 0 else 0.0,
        )
        for h in merged.values()
    ]
    holdings.sort(key=lambda h: h.value, reverse=True)
    return holdings

File: lib/thirteenf.ts
/**
 * SEC EDGAR 13F holdings fetcher — institutional "smart money" tracker.
 *
 * Pulls a fund manager's most recent 13F-HR filing (the quarterly disclosure
 * of long US-equity holdings) directly from EDGAR. 13F data only changes once
 * a quarter, so the EDGAR responses are cached for 6h via `next.revalidate`.
 */

const UA =
  process.env.EDGAR_USER_AGENT || "ai-trading-agent/1.0 (contact@example.com)";
const REVALIDATE = 6 * 60 * 60; // 6 hours — 13F data updates quarterly

export interface Manager {
  name: string;
  cik: string; // 10-digit zero-padded
}

export interface Holding {
  issuer: string;
  cusip: string;
  value: number; // USD, as reported in the filing
  shares: number;
  pct: number; // share of the manager's total 13F portfolio (0..1)
}

export interface Report {
  manager: string;
  cik: string;
  periodOfReport: string | null; // e.g. "2026-03-31"
  filedDate: string | null;
  totalValue: number;
  holdings: Holding[]; // merged by issuer, sorted by value desc
  error?: string;
}

/** Widely-followed institutional managers, by SEC CIK. */
export const MANAGERS: Manager[] = [
  { name: "Berkshire Hathaway — Warren Buffett", cik: "0001067983" },
  { name: "Scion Asset Management — Michael Burry", cik: "0001649339" },
  { name: "Pershing Square — Bill Ackman", cik: "0001336528" },
];

interface SubmissionsRecent {
  form?: string[];
  accessionNumber?: string[];
  reportDate?: string[];
  filingDate?: string[];
}

async function edgarJson<T>(url: string): Promise<T> {
  const res = await fetch(url, {
    headers: { "User-Agent": UA, Accept: "application/json" },
    next: { revalidate: REVALIDATE },
  });
  if (!res.ok) throw new Error(`EDGAR ${res.status}`);
  return (await res.json()) as T;
}

async function edgarText(url: string): Promise<string> {
  const res = await fetch(url, {
    headers: { "User-Agent": UA },
    next: { revalidate: REVALIDATE },
  });
  if (!res.ok) throw new Error(`EDGAR ${res.status}`);
  return res.text();
}

const ENTITIES: Record<string, string> = {
  "&amp;": "&",
  "&lt;": "<",
  "&gt;": ">",
  "&quot;": '"',
  "&#39;": "'",
  "&apos;": "'",
};

/** Read one tag's text from an XML fragment (namespace-prefix tolerant). */
function tag(xml: string, name: string): string | null {
  const m = xml.match(new RegExp(`<(?:\\w+:)?${name}\\b[^>]*>([^<]*)<`, "i"));
  if (!m) return null;
  return m[1]
    .trim()
    .replace(/&(amp|lt|gt|quot|#39|apos);/g, (e) => ENTITIES[e] ?? e);
}

export async function getLatest13F(manager: Manager): Promise<Report> {
  const base: Report = {
    manager: manager.name,
    cik: manager.cik,
    periodOfReport: null,
    filedDate: null,
    totalValue: 0,
    holdings: [],
  };
  try {
    const subs = await edgarJson<{ filings?: { recent?: SubmissionsRecent } }>(
      `https://data.sec.gov/submissions/CIK${manager.cik}.json`,
    );
    const recent = subs.filings?.recent ?? {};
    const forms = recent.form ?? [];
    const idx = forms.findIndex((f) => f === "13F-HR" || f === "13F-HR/A");
    if (idx < 0) return { ...base, error: "no 13F-HR filing found" };

    const accession = recent.accessionNumber?.[idx] ?? "";
    const periodOfReport = recent.reportDate?.[idx] ?? null;
    const filedDate = recent.filingDate?.[idx] ?? null;
    const cikInt = String(Number(manager.cik));
    const dir = `https://www.sec.gov/Archives/edgar/data/${cikInt}/${accession.replace(/-/g, "")}`;

    const index = await edgarJson<{
      directory?: { item?: { name: string }[] };
    }>(`${dir}/index.json`);
    const infoFile = (index.directory?.item ?? []).find((it) =>
      /(?:info|information)table.*\.xml$/i.test(it.name),
    );
    if (!infoFile) {
      return {
        ...base,
        periodOfReport,
        filedDate,
        error: "no info table in filing",
      };
    }

    const xml = await edgarText(`${dir}/${infoFile.name}`);
    const blocks =
      xml.match(/<(?:\w+:)?infoTable\b[\s\S]*?<\/(?:\w+:)?infoTable>/gi) ?? [];

    // 13F lists each lot separately; merge by issuer (keyed on CUSIP).
    const merged = new Map<string, Holding>();
    for (const block of blocks) {
      const value = Number(tag(block, "value") ?? "0");
      if (!(value > 0)) continue;
      const issuer = tag(block, "nameOfIssuer") ?? "—";
      const cusip = tag(block, "cusip") ?? "";
      const shares = Number(tag(block, "sshPrnamt") ?? "0");
      const key = cusip || issuer;
      const prev = merged.get(key);
      if (prev) {
        prev.value += value;
        prev.shares += shares;
      } else {
        merged.set(key, { issuer, cusip, value, shares, pct: 0 });
      }
    }

    const holdings = [...merged.values()].sort((a, b) => b.value - a.value);
    const totalValue = holdings.reduce((s, h) => s + h.value, 0);
    for (const h of holdings) h.pct = totalValue > 0 ? h.value / totalValue : 0;

    return { ...base, periodOfReport, filedDate, totalValue, holdings };
  } catch (e) {
    return { ...base, error: e instanceof Error ? e.message : String(e) };
  }
}
